homework-3

MrFa1n opened the pull request:
- add filter to posts by length
- update modal to compound components pattern
- add hoc withLoading
- add less/more btn to post
- optimization post view

MrFa1n committed:
feat: enhance UI components with improved styles, add Button and Modal functionality, and update layout structure

MrFa1n committed:
feat:
- add filter to posts by length
- update modal to compound components pattern
- add hoc withLoading
- add less/more btn to post
- optimization post view

maksimstorazhau-dot commented on src/app/App.tsx:
```diff
@@ -0,0 +1,20 @@
+  const [posts] = useState(mockPosts);
+  const [loading] = useState(false);
+
+  const PostListWithLoading = withLoading(PostList);
```
Лучше это делать вне компонента, иначе каждый рендер = новый компонент

maksimstorazhau-dot commented on src/entities/post/ui/PostCard.tsx:
```diff
@@ -0,0 +1,13 @@
+import { CommentsList } from "@/widgets/CommentsList/ui/CommentsList";
```
Тут нарушение FSD, лучше через пропсы прокидывать

maksimstorazhau-dot commented on src/features/PostLengthFilter/ui/PostLengthFilter.tsx:
```diff
@@ -0,0 +1,31 @@
+  onFilter: (length: number) => void;
+}
+
+export const PostLengthFilter = ({ onFilter }: Props) => {
```
Оптимизация с useCallback в PostList работать не будет без мемо

maksimstorazhau-dot commented on src/features/PostLengthFilter/ui/PostLengthFilter.tsx:
```diff
@@ -0,0 +1,31 @@
+      <input
+        className={styles.filterInput}
+        type="number"
+        value={length}
```
Не вижу особого смысла контролировать инпут  это лишь создаёт лишние перерисовки. Лучше просто использовать defaultValue={50}

maksimstorazhau-dot commented on src/features/PostLengthFilter/ui/PostLengthFilter.tsx:
```diff
@@ -0,0 +1,31 @@
+      onFilter(val);
+    },
+    [onFilter]
+  );
```
Тут еще лучше добавлять дебаунс с минимальной задержкой

maksimstorazhau-dot commented on src/features/ThemeSwitcher/ui/ThemeSwitcher.tsx:
```diff
@@ -0,0 +1,44 @@
+          <line x1="4.22" y1="19.78" x2="6.34" y2="17.66" />
+          <line x1="17.66" y1="6.34" x2="19.78" y2="4.22" />
+        </g>
+      </svg>
```
Не очень читаемо  лучше вынести иконки в отдельные компоненты, например: Moon и Sun

maksimstorazhau-dot commented on src/shared/ui/Button/Button.tsx:
```diff
@@ -0,0 +1,48 @@
+  type?: "button" | "submit" | "reset";
+};
+
+export const Button: React.FC<ButtonProps> = ({
```
Тут было бы неплохо использовать мемо

maksimstorazhau-dot commented on src/shared/ui/Modal/Modal.tsx:
```diff
@@ -0,0 +1,82 @@
+        ></path>{" "}
+      </g>
+    </svg>
+  );
```
Так же лучше иконку вынести в отдельный компонент

maksimstorazhau-dot commented on src/shared/ui/Modal/Modal.tsx:
```diff
@@ -0,0 +1,82 @@
+        </Button>
+      </div>
+    </div>,
+    document.body
```
Лучше в html добавить тег c id modal-root и рендерить туда

maksimstorazhau-dot commented on src/widgets/CommentsList/ui/CommentsList.tsx:
```diff
@@ -0,0 +1,131 @@
+
+export const CommentsList = ({ comments }: CommentsListProps) => {
+  const [collapsed, setCollapsed] = useState(true);
+  const toggle = useCallback(() => setCollapsed((prev) => !prev), []);
```
useCallback не работает, проблема в Button

maksimstorazhau-dot commented on src/widgets/CommentsList/ui/CommentsList.tsx:
```diff
@@ -0,0 +1,131 @@
+        fill="#0000000"
+        version="1.1"
+        id="Capa_1"
+        xmlns="http://www.w3.org/2000/svg"
+        viewBox="-3.17 -3.17 37.99 37.99"
+        stroke="#0000000"
+      >
+        <g
+          id="SVGRepo_bgCarrier"
+          stroke-width="0"
+          transform="translate(0,0), scale(1)"
+        >
+          <rect
+            x="-3.17"
+            y="-3.17"
+            width="37.99"
+            height="37.99"
+            rx="18.995"
+            fill="#ffffff"
+          ></rect>
+        </g>
+        <g
+          id="SVGRepo_tracerCarrier"
+          stroke-linecap="round"
+          stroke-linejoin="round"
+          stroke="#CCCCCC"
+          stroke-width="0.1266"
+        ></g>
+        <g id="SVGRepo_iconCarrier">
+          {" "}
+          <g>
+            {" "}
+            <g>
+              {" "}
+              <path d="M26.734,26.547l-1.395-8.357c-0.366-2.189-2.261-3.794-4.479-3.794h-2.273c2.754-1.098,4.703-3.784,4.703-6.931 C23.29,3.341,19.948,0,15.826,0S8.362,3.342,8.362,7.464c0,3.147,1.95,5.833,4.705,6.931h-2.276c-2.22,0-4.114,1.604-4.479,3.794 l-1.395,8.357c-0.212,1.271,0.146,2.569,0.979,3.553S7.951,31.65,9.24,31.65h13.172c1.289,0,2.512-0.567,3.345-1.551 C26.588,29.117,26.946,27.817,2 ...
+            </g>{" "}
+          </g>{" "}
+        </g>
+      </svg>
+    );
```
В отдельный компонент

maksimstorazhau-dot commented on src/widgets/CommentsList/ui/CommentsList.tsx:
```diff
@@ -0,0 +1,131 @@
+                <p className={styles.commentName}>{c.name}</p>
+                <p className={styles.commentBody}>{c.text}</p>
+              </div>
+            </p>
```
Тут можно вынести в entities/Comment

maksimstorazhau-dot commented on src/widgets/LayoutHeader/Header.tsx:
```diff
@@ -0,0 +1,44 @@
+      <div className={styles.content}>
+        <div className={styles.logo}>FakePress</div>
+        <div className={styles.navigation}>
+          <Button
+            onClick={() => {
+              setIsOpen(true);
+            }}
+            variant="glassBlue"
+            shape="pill"
+          >
+            О проекте
+          </Button>
```
При изменении isOpen компонент будет перерисовываться. Лучше обернуть обработчик в useCallback, и в целом обработчики лучше не объявлять прямо в JSX

maksimstorazhau-dot commented on src/widgets/LayoutHeader/Header.tsx:
```diff
@@ -0,0 +1,44 @@
+          omnis temporibus cum dolorem nihil. Qui neque iure rem aut soluta
+          dolorum.
+        </Modal.Body>
+      </Modal>
```
Я бы рядом с Header.tsx создал файл HeaderModal.tsx и вынес модалку туда

maksimstorazhau-dot commented on src/widgets/PostList/PostList.tsx:
```diff
@@ -0,0 +1,38 @@
+    [posts, maxLength]
+  );
+
+  const handleFilter = useCallback(
+    (length: number) => setMaxLength(length),
+    []
+  );
```
useCallback не работает проблема в PostLengthFilter

MrFa1n committed:
- add modal root container
- add new icon components
- add comment card entities
- optimize memo/callback components
- optimize button component
- optimize render for withLoading component

MrFa1n committed:
Merge remote-tracking branch 'origin/develop' into feature/modal-filter-comments-optimizations
